multidim-interop/rust/src/bin/testplan_0490.rs

```diff
@@ -0,0 +1,175 @@
+use testground::client::Client as TGClient;
+use testplan::{run_ping, PingSwarm, TransportKind};
+
+#[async_std::main]
+async fn main() -> Result<()> {
+    let local_key = identity::Keypair::generate_ed25519();
+    let local_peer_id = PeerId::from(local_key.public());
+
+    let client = TGClient::new_and_init().await.unwrap();
+
+    let transport_param: String = client
+        .run_parameters()
+        .test_instance_params
+        .get("transport")
+        .unwrap()
+        .parse()
+        .unwrap();
+
+    let secure_channel_param: String = client
+        .run_parameters()
+        .test_instance_params
+        .get("security")
+        .unwrap()
+        .parse()
+        .unwrap();
+
+    let muxer_param: String = client
+        .run_parameters()
+        .test_instance_params
+        .get("muxer")
+        .unwrap()
+        .parse()
+        .unwrap();
+
+    // Lots of duplication because I, Marco, couldn't figure out how to make the type system happy.
+    let boxed_transport = match transport_param.as_str() {
+        "tcp" => {
+            let builder = TcpTransport::new(GenTcpConfig::new())
+                .upgrade(libp2p::core::upgrade::Version::V1Lazy);
+
+            let authenticated = match secure_channel_param.as_str() {
+                "noise" => builder
+                    .authenticate(libp2pv0490::noise::NoiseAuthenticated::xx(&local_key).unwrap()),
+                _ => panic!("Unsupported secure channel"),
+            };
+
+            match muxer_param.as_str() {
+                "yamux" => authenticated
+                    .multiplex(yamux::YamuxConfig::default())
+                    .timeout(Duration::from_secs(5))
+                    .boxed(),
+                "mplex" => authenticated
+                    .multiplex(mplex::MplexConfig::default())
+                    .timeout(Duration::from_secs(5))
+                    .boxed(),
+                _ => panic!("Unsupported muxer"),
+            }
+        }
+        "ws" => {
+            let builder = WsConfig::new(TcpTransport::new(GenTcpConfig::new()))
+                .upgrade(libp2p::core::upgrade::Version::V1Lazy);
+
+            let authenticated = match secure_channel_param.as_str() {
+                "noise" => builder
+                    .authenticate(libp2pv0490::noise::NoiseAuthenticated::xx(&local_key).unwrap()),
+                _ => panic!("Unsupported secure channel"),
+            };
+
+            match muxer_param.as_str() {
+                "yamux" => authenticated
+                    .multiplex(yamux::YamuxConfig::default())
+                    .timeout(Duration::from_secs(5))
+                    .boxed(),
+                "mplex" => authenticated
+                    .multiplex(mplex::MplexConfig::default())
+                    .timeout(Duration::from_secs(5))
+                    .boxed(),
+                _ => panic!("Unsupported muxer"),
+            }
+        }
+        _ => panic!("Unsupported"),
+    };
```

Comment:
The question is, do we want to have a node that _only_ speaks the requested configuration or should it just be compatible with it.

Even in the first case, the easier option here is probably to use `OptionalTransport` and `OptionalUpgrade`. That allows you move the conditional code into a value that you can compose as if it were unconditional!

Reply:
We want something that _only_ speaks the requested configuration. That way we are sure we are testing the requested configuration and not accidentally making something else work.

This code is definitely hacked together. I'll take a look at the OptionalTransport, thanks!

Reply:
@jxs Do you think you can help me clean this up? You'll certainly do this faster than me.

Reply:
I switched to using an EitherUpgrade